packages/flutter/lib/src/material/ink_sparkle.dart

```diff
@@ -0,0 +1,3661 @@
+      canvas.clipPath(
+          customBorder.getOuterPath(rect, textDirection: textDirection));
+    } else if (borderRadius != BorderRadius.zero) {
+      canvas.clipRRect(RRect.fromRectAndCorners(
+        rect,
+        topLeft: borderRadius.topLeft,
+        topRight: borderRadius.topRight,
+        bottomLeft: borderRadius.bottomLeft,
+        bottomRight: borderRadius.bottomRight,
+      ));
+    } else {
+      canvas.clipRect(rect);
+    }
+  }
+}
+
+class _InkSparkleFactory extends InteractiveInkFeatureFactory {
+  const _InkSparkleFactory() : turbulenceSeed = null;
+
+  const _InkSparkleFactory.constantTurbulenceSeed() : turbulenceSeed = 1337.0;
+
+  // TODO(clocksmith): Update this once shaders are precompiled.
+  static void compileShaderIfNeccessary() {
+    if (!_initCalled) {
+      FragmentShaderManager.inkSparkle().then((FragmentShaderManager manager) {
+        _shaderManager = manager;
+      });
+      _initCalled = true;
+    }
+  }
+  static bool _initCalled = false;
+  static FragmentShaderManager? _shaderManager;
+
+  final double? turbulenceSeed;
+
+  @override
+  InteractiveInkFeature create({
+    required MaterialInkController controller,
+    required RenderBox referenceBox,
+    required ui.Offset position,
+    required ui.Color color,
+    required ui.TextDirection textDirection,
+    bool containedInkWell = false,
+    RectCallback? rectCallback,
+    BorderRadius? borderRadius,
+    ShapeBorder? customBorder,
+    double? radius,
+    ui.VoidCallback? onRemoved,
+  }) {
+    return InkSparkle(
+      controller: controller,
+      referenceBox: referenceBox,
+      position: position,
+      color: color,
+      textDirection: textDirection,
+      containedInkWell: containedInkWell,
+      rectCallback: rectCallback,
+      borderRadius: borderRadius,
+      customBorder: customBorder,
+      radius: radius,
+      onRemoved: onRemoved,
+      turbulenceSeed: turbulenceSeed,
+    );
+  }
+}
+
+RectCallback? _getClipCallback(
+  RenderBox referenceBox,
+  bool containedInkWell,
+  RectCallback? rectCallback,
+) {
+  if (rectCallback != null) {
+    assert(containedInkWell);
+    return rectCallback;
+  }
+  if (containedInkWell) {
+    return () => Offset.zero & referenceBox.size;
+  }
+  return null;
+}
+
+double _getTargetRadius(
+  RenderBox referenceBox,
+  bool containedInkWell,
+  RectCallback? rectCallback,
+  Offset position,
+) {
+  final Size size =
+      rectCallback != null ? rectCallback().size : referenceBox.size;
+  final double d1 = size.bottomRight(Offset.zero).distance;
+  final double d2 =
+      (size.topRight(Offset.zero) - size.bottomLeft(Offset.zero)).distance;
+  return math.max(d1, d2) / 2.0;
+}
+
+// The code below is generated by the package: fragment_shader_manager.
+// See:
+// - [https://github.com/material-components/material-components-flutter-experimental/tree/fragment-shader-manager/fragment_shader_manager]
+
+// TODO(REVIEWERS): Move to sep file?
```

Comment:
I'm not that familiar with our shader integration system, but this seems like it might be better to have in a separate file to make it easier to read the surrounding code.

Reply:
Since there are no sibling files with .g.dart, I decided to keep here, but more docs explaining where it comes from